timemachine/fe/single_topology.py

```diff
@@ -1638,94 +1690,47 @@ def setup_intermediate_state(self, lamb) -> VacuumSystem:
-                lambda_min=bonds_min,
-                lambda_max=bonds_max,
-            ),
-        )
-
-        angle = self._setup_intermediate_bonded_term(
-            src_system.angle,
-            dst_system.angle,
-            lamb,
-            interpolate.align_harmonic_angle_idxs_and_params,
-            partial(
-                interpolate_harmonic_angle_params,
-                k_min=0.05,  # ~ BOLTZ * (300 K) / (2 * pi)^2
-                lambda_min=angles_min,
-                lambda_max=angles_max,
-            ),
-        )
-
-        assert src_system.torsion
-        assert dst_system.torsion
-        torsion = self._setup_intermediate_bonded_term(
-            src_system.torsion,
-            dst_system.torsion,
-            lamb,
-            interpolate.align_torsion_idxs_and_params,
-            partial(interpolate_periodic_torsion_params, lambda_min=torsions_min, lambda_max=torsions_max),
-        )
-
-        nonbonded = self._setup_intermediate_nonbonded_term(
-            src_system.nonbonded,
-            dst_system.nonbonded,
-            lamb,
-            interpolate.align_nonbonded_idxs_and_params,
-            interpolate.linear_interpolation,
-        )
-
-        assert src_system.chiral_atom
-        assert dst_system.chiral_atom
-
-        assert len(set(tuple(x) for x in src_system.chiral_atom.potential.idxs)) == len(
-            src_system.chiral_atom.potential.idxs
-        )
-        assert len(set(tuple(x) for x in dst_system.chiral_atom.potential.idxs)) == len(
-            dst_system.chiral_atom.potential.idxs
-        )
-
-        chiral_atom = self._setup_intermediate_bonded_term(
-            src_system.chiral_atom,
-            dst_system.chiral_atom,
-            lamb,
-            interpolate.align_chiral_atom_idxs_and_params,
-            partial(
-                interpolate_harmonic_force_constant,
-                k_min=0.025,
-                lambda_min=chiral_atoms_min,
-                lambda_max=chiral_atoms_max,
-            ),
-        )
+        bond = self.align_and_interpolate_bonds(lamb)
+        angle = self.align_and_interpolate_angles(lamb)
+        chiral_atom = self.align_and_interpolate_chiral_atoms(lamb)
+        torsion = self.align_and_interpolate_torsions(lamb)
+        nonbonded = self.align_and_interpolate_intramolecular_nonbonded(lamb)
 
         assert src_system.chiral_bond
         assert dst_system.chiral_bond
+        # (ytz): dead code, not actually used in production
         chiral_bond = self._setup_intermediate_chiral_bond_term(
             src_system.chiral_bond,
             dst_system.chiral_bond,
             lamb,
             interpolate.linear_interpolation,
         )
 
-        return VacuumSystem(bond, angle, torsion, nonbonded, chiral_atom, chiral_bond)
+        vs = VacuumSystem(bond, angle, torsion, nonbonded, chiral_atom, chiral_bond)
+        assert_bonds_defined_for_chiral_volumes(vs, DEFAULT_BOND_IS_PRESENT_K)
```

Comment:
@maxentile I added a function to verify that bonds are present for chiral volumes in the `setup_intermediate_state()` call - however, I'm running into a problem that this is incompatible with `jax.jit`

@mcwitt - any ideas? maybe checkify?

``` python
    def assert_bonds_defined_for_chiral_volumes(vs: VacuumSystem, bond_k_min: float):
        """
        Assert that bonds defined for every chiral volume is present and has a force constant greater than bond_k_min
        """
        bonds_present = set()
        for idxs, (bond_k,_) in zip(vs.bond.potential.idxs, vs.bond.params):
>           if bond_k > bond_k_min:
E           jax.errors.TracerBoolConversionError: Attempted boolean conversion of traced array with shape bool[]..
E           The error occurred while tracing the function U at /home/yzhao/Code/timemachine/tests/test_single_topology.py:641 for jit. This concrete value was not available in Python because it depends on the value of the argument lam.
E           See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerBoolConversionError
```

Reply:
I think we have a pattern in a few other places where we add a `verify` argument (defaulting to true), and guard the assertions so they are only executed if `verify == True`. Then I think `jit(partial(f, verify=False))` or something like `jit(f, static_argnames=("verify",))` (and then passing `verify=False`) can be supported?

Reply:
Looking again, I only see one example of this in the current code (and the flag is called `runtime_validate`, not `verify`): https://github.com/proteneer/timemachine/blob/be34fbf458b54de5454564b67aeb1d70450a1b3f/timemachine/potentials/nonbonded.py#L228

I think `checkify` would work too, but the downside is I think it complicates usage without jax?

Reply:
I'd also be OK with breaking jax transformability in the short term; I don't think this is used in production, and `jit` probably isn't the right long-term solution for SingleTopology slowness

Reply:
I'm going to punt on this for now, and just do the verification *outside* of the calling function.